src/contracts/middleware/Middleware.sol

```diff
@@ -306,13 +307,55 @@ contract Middleware is
         bytes calldata /* checkData */
     ) external view override returns (bool upkeepNeeded, bytes memory performData) {
         uint48 epoch = getCurrentEpoch();
-        bytes32[] memory sortedKeys = IOBaseMiddlewareReader(address(this)).sortOperatorsByPower(epoch);
+        uint48 currentEpochStartTs = IOBaseMiddlewareReader(address(this)).getEpochStart(epoch);
+
         StorageMiddleware storage $ = _getMiddlewareStorage();
+        StorageMiddlewareCache storage cache = _getMiddlewareStorageCache();
+
+        address[] memory activeOperators = _activeOperators();
+        uint256 activeOperatorsLength = activeOperators.length;
+        if (activeOperatorsLength == 0) {
+            // No active operators, no upkeep needed
+            return (false, hex"");
+        }
+
+        uint256 cachedCurrentEpochValidatorsLength = cache.epochToValidatorsData[epoch].length;
+        uint256 diffLength = activeOperatorsLength - cachedCurrentEpochValidatorsLength;
+
+        // Check if cache is still not filled with the current epoch validators
+        if (diffLength > 0) {
+            uint256 maxNumOperatorsToCheck = Math.min(diffLength, MAX_OPERATORS_TO_PROCESS);
+
+            ValidatorData[] memory validatorsData = new ValidatorData[](maxNumOperatorsToCheck);
+
+            // Populate validatorsData with the new operators' keys and their powers
+            // It gets encoded to be used in performUpkeep
+            for (
+                uint256 i = cachedCurrentEpochValidatorsLength;
+                i < cachedCurrentEpochValidatorsLength + maxNumOperatorsToCheck && i < activeOperatorsLength;
+                i++
```

Comment:
Not sure if this is efficient on 0.8.25, or if it's still better to do the `unchecked i++`

Reply:
Found this on CHANGELOG for 0.8.22. So unchecked is probably not needed anymore. Will try to test the gas usage today.
> Code Generator: Remove redundant overflow checks of certain for loops when the counter variable cannot overflow.